experimental: expose source offset to Listener

pelletier opened the pull request:
This change provides a user of `experimental.StackIterator` the ability to retrieve the offset in the Code section of each function call in the frame of the stack. It enables capabilities such as source-mapping to be implemented by the user outside of wazero (removing the need for #1403).

It works by retrieving the return address of each stack frame and using the `irOperationSourceOffsetsInWasmBinary` attached to the function associated with that stack frame.

Though it works in my tests and on the few programs I used, I feel quite out of my depth here; I'd love some guidance and feedback! :) 

Question 1 – This source offsets mapping is only produced by the compiler when DWARF sections are present in the module. In practice, it's a reasonable limitation for us since we currently are only interested in using this information to correlate with DWARF data anyway. However, we may want to look at other applications in the future. It also meant that I had to provide a minimal DWARF info section for `dwarf/debug` not to trip up and the compiler to emit the offsets map. Would you be interested in adding either a flag to enable the mapping generation regardless of the presence of DWARF sections?

Question 2 – because the stack iterator is used in the`Before` listener hook, the topmost frame of the call stack is always the called function that triggered the listener. As such, I think the source offset should be the offset of the first byte of that function in the code section. In that case, I have observed that the pc is just lower than the first entry of `irOperationSourceOffsetsInWasmBinary` (hence the change in `getSourceOffsetInWasmBinary`). I am struggling to understand why? Returning the first entry of that collection seems to give the expected results, but I don't rationalize it.

I appreciate your consideration!

codefromthecrypt reviewed (commented):
summary comments as this morning is a meeting morning.

This sort of feature should have a test in the experimental package to show its use, even minimally.

Also some bikeshedding about returning -1 instead of zero, for example for host functions..

and the length of the code section I think will not be 64bits :) sounds like something to quote and be wrong 20 years later ;)

The comment on the exposed thing should continue to say how it is used. `e.g. to allow debuggers to map this function to its source`

I'll leave other comments to others, then later chime in more if needed.

Meanwhile, cool stuff!

pelletier commented:
Thanks for the review! 

Expanded the documentation for `SourceOffset` and added it to the existing `StackIterator` example (and fixed it along the way...)

> Also some bikeshedding about returning -1 instead of zero, for example for host functions.
> and the length of the code section I think will not be 64bits :)

That's fair. `getSourceOffsetInWasmBinary` (and equivalent in the interpreter) uses `uint64`. It suffers from the same issue of returning 0 if not found (which I think would be a valid offset if called with the PC at the beginning very first function of the Code section?) and using a rather large type. I opted to use `uint64` for SourceOffset for consistency, as it seems to be the type used for offsets in the Code section in Wazero.

To signal a "not found" case, I thought about either using another sentinel value (like `MaxUint64`) or returning an additional `error` value in that case but eventually decided to lean into zero-values for less ceremony. Aside from that, I don't have much argument either way.

I'm happy to change the signature of `SourceOffset` to any of the above and do the appropriate casts, and/or change `getSourceOffsetInWasmBinary` & friends to have them not use `uint64` either. Let me know your preference!

codefromthecrypt commented:
I personally prefer -1 to zero for saying not found (e.g. strings.Index) as it feels more idiomatic in Go, I'm not aware of a lot of use of Max or zero for the same thing. It is a bike shed, and to get through them, usually blaming decision on another precedent is the easy way out. I would blame other code that look for an index of something in go returning -1 as the reason.

mathetake commented on internal/engine/compiler/engine.go:
```diff
@@ -880,6 +880,9 @@ func (f *function) getSourceOffsetInWasmBinary(pc uint64) uint64 {
 		}
 		return srcMap.irOperationOffsetsInNativeBinary[i] >= pcOffsetInNativeBinary
 	})
+	if index == 0 && len(srcMap.irOperationSourceOffsetsInWasmBinary) > 0 {
```
I think I need comment on this for the future

codefromthecrypt replied:
personally I would ask "why is this done (only) on index zero" and want the comment to answer that. Up to others what the good comment would be

evacchi replied:
I wonder if it's worth adding a `func hasSourceOffsetsInWasmBinary() bool { return len(irOperationSourceOffsetsInWasmBinary) != 0 }` (maybe not in this PR)

pelletier replied:
Added a comment to clarify.

mathetake reviewed (commented):
looks good to me. Requesting review from @evacchi as well and delegating to @codefromthecrypt and edo for the final approval. Good work!

codefromthecrypt commented on experimental/listener.go:
```diff
@@ -13,14 +13,22 @@ import (
 	FunctionDefinition() api.FunctionDefinition
+	// SourceOffset computes the offset in the module Code section where the
+	// call occured (translated for native functions), or the beginning of
+	// function for the top of the stack. Returns 0 if the source offset
```
@ncruces have any 👨‍🎨 on the -1 vs zero thing for source offset not available. I know we recently chatted on tinygo (and otherwise) using zero as nil though technically wasm can start from zero.

I think basically nothing will start at line zero in the source file, is why we can use zero here. (it isn't about the binary rather the %.c %.rs etc "the source"

ncruces replied:
Yes, I don't think what we chatted about applies here.

TLDR it was: linear memory address zero in C (and probably TinyGo) can't be dereferenced, so I choose to panic when the host gets a zero pointer it's expected to dereference (although the host _can_ read/write address zero).

ncruces replied:
As long as 0 is not a valid offset and given this is unsigned, zero seams reasonable.

codefromthecrypt replied:
apples and oranges, but what's common (fruit?) is we are banishing a zero offset which isn't usually valid.

So, the simile is like memory offset zero is almost always invalid, source offset zero would be pretty invalid.

nerdy side bar:

As far as I can think of, the only languages which could use either are %.wat and an edge case of that, also. For example, text format can use memory offset zero, and also strangely, you can omit `module` and have a function as a top-level element in the file.. otoh I doubt if source-mapped you'd end up with index zero as it would not be in the body part ;) https://www.w3.org/TR/wasm-core-1/#abbreviations%E2%91%A0%E2%91%A5

codefromthecrypt commented on internal/engine/interpreter/interpreter.go:
```diff
@@ -261,6 +264,16 @@ func (si *stackIterator) FunctionDefinition() api.FunctionDefinition {
 	return si.fn.definition()
 }
 
+// SourceOffset implements experimental.StackIterator.
```
nit: to avoid introducing style 3 or 4 of this boilerplate ;)

```suggestion
// SourceOffset implements the same method as documented on experimental.StackIterator.
```

codefromthecrypt commented on internal/engine/compiler/engine.go:
```diff
@@ -1128,13 +1133,25 @@ func (si *stackIterator) Next() bool {
 	return si.fn != nil
 }
 
+// SourceOffset implements experimental.StackIterator.
```
```suggestion
// SourceOffset implements the same method as documented on experimental.StackIterator.
``````

codefromthecrypt commented on internal/engine/compiler/engine.go:
```diff
@@ -1128,13 +1133,25 @@ func (si *stackIterator) Next() bool {
+	p := si.fn.parent
+
+	if len(p.sourceOffsetMap.irOperationSourceOffsetsInWasmBinary) == 0 {
+		return 0
```
```suggestion
		return 0 // source isn't available
```

codefromthecrypt reviewed (approved):
looks to me there's zero perf impact unless stackiterator is already in use. cool

nits only, so ready on green after the comment request is in!

evacchi reviewed (approved):
I don't have much to add besides what @mathetake and @codefromthecrypt already indicated :)

evacchi commented:
ps: obviously there are still a few failures on CI that should be addressed

pelletier commented:
Thank you everyone for the reviews! I have addressed all the comments (unless I misunderstood the -1 vs 0 decision).

codefromthecrypt commented on experimental/listener_example_test.go:
```diff
@@ -89,32 +89,44 @@ func Example_stackIterator() {
-	// function: fn2 args [4]
+	// function: fn0
+	// 	parameters: [1 2 3]
+	// 	source offset: 1234
```
👍

codefromthecrypt commented on internal/engine/compiler/engine.go:
```diff
@@ -1116,7 +1124,7 @@ func (si *stackIterator) clear() {
 }
 
-// Next implements experimental.StackIterator.
+// Next implements the same method as documtend on experimental.StackIterator.
```
```suggestion
// Next implements the same method as documented on experimental.StackIterator.
```

codefromthecrypt commented:
nicely done! I'll polish the typo post merge in another change
